h5netcdf/core.py

```diff
@@ -125,28 +132,61 @@ def shuffle(self):
 
 NOT_A_VARIABLE = b'This is a netCDF dimension but not a netCDF variable.'
 
+class lazy_objects(Mapping):
+    def __init__(self, object_cls):
+        self._objects = list()
+        self._loaded_objects = dict()
+        self._object_cls = object_cls
+        return
+
+    def set(self, name, object):
+        self._objects.append(name)
+        self._loaded_objects[name] = object
+        return
+
+    def add(self, name):
+        self._objects.append(name)
+        return
+
+    def __iter__(self):
+        for name in self._objects:
+            yield name
+
+    def __len__(self):
+        return len(self._objects)
+
+    def __getitem__(self, key):
+        if not key in self._objects:
+            raise KeyError('Object {0} is not avaialble.'.format(key))
+        elif key in self._loaded_objects.keys():
```

Comment:
`elif key in self._loaded_objects:` avoids constructing a list for `.keys()` on Python 2.